doc/developer-guide/internal-libraries/Extendible.en.rst

```diff
@@ -0,0 +1,143 @@
+* Extendible<DerivedType> - allocates block of memory, uses FieldId or schema to access slices of memory.
+
+
+During system init, code and plugins can add fields to the Extendible's schema. This will update the `Memory Layout`_ of the schema, and the memory offsets of all fields. The schema does not know the FieldType, but it stores the byte size and creates lambdas of the type's constructor and destructor. And to avoid corruption, the code asserts that no instances are in use when adding fields.
+
+.. code-block:: cpp
+
+   ExtendibleExample:FieldId<ATOMIC,int> fld_my_int;
+
+   void PluginInit() {
+     fld_my_int = ExtendibleExample::schema.addField("my_plugin_int");
+   }
+
+
+When an Extendible derived class is instantiated, new() will allocate a block of memory for the derived class and all added fields. There is zero memory overhead per instance, unless using :ref:ACIDPTR field access type.
+
+.. code-block:: cpp
+
+   ExtendibleExample* alloc_example() {
+     return new ExtendibleExample();
+   }
+
+Memory Layout
+-------------
+One block of memory is allocated per Extendible, which include all member variables and appended fields.
+Within the block, memory is arranged in the following order:
+
+#. Derived members (+padding align next field)
+#. Fields (largest to smallest)
+#. Packed Bits
+
++-------------------------------------------+
+|Extendible::new()                          |
++-----------------+-------------------------+
+|ExtendibleExample|Fields                   |
++-----------------+-------------------------+
+|int real_member  |atomic<int> my_plugin_int|
++-----------------+-------------------------+
+
+Strongly Typed
+--------------
+Extendible was written for efficient thread safety. When a field is added to the Extendible type, data type and access type part of the addField template, resulting in a strongly typed FieldId, and then will be implicitly enforced through every access of that field.
+
+.. code-block:: cpp
+
+   class Food : Extendible<Food> {}
+   auto fld_food_weight = Food.schema.addField<STATIC,float>("weight");
+   auto fld_expr_date = Food.schema.addField<STATIC,time_t>("expire date");
+   Food banana;
+
+   class Car : Extendible<Car> {}
+   auto fld_max_speed = Car.schema.addField<STATIC,float>("max_speed");
+   auto fld_car_weight = Car.schema.addField<STATIC,float>("weight"); // 'weight' is unique within 'Car'
+   Car camry;
+
+   // Common user errors
+
+   float expire_date = banana.get(fld_expr_date);
+   //^^^                                              Compile error: cannot convert time_t to float
+   float speed = banana.get(fld_max_speed);
+   //                       ^^^^^^^^^^^^^             Compile error: fld_max_speed is not of type Extendible<Food>::FieldId
+   float weight = camry.get(fld_food_weight);
+   //                       ^^^^^^^^^^^^^^^           Compile error: fld_food_weight is not of type Extendible<Car>::FieldId
+
```

Comment:
Have you looked at what it would take to integrate this with `std::get`? E.g. so `std::get<fld_expr_date>(banana)` works as expected. That would make it more `std::tuple` like, which might be good.

Reply:
That is a good idea.

Reply:
* This is the first I see `addField` mentioned anywhere. In fact, I can't find the documentation for `Schema` either. It would be nice to have a "Reference" like [MemArena](https://github.com/apache/trafficserver/blob/master/doc/developer-guide/internal-libraries/MemArena.en.rst#reference).

* This example has calls to `addField`, but I don't know what the signatures of it are. I can see from the example that it accepts a string literal, but what becomes of this string? One example `"expire date"` has a space in in, so I figure these don't need to be valid symbols. What are they for?

* I see the following declarations of `Extendible::Schema::addField`:
   * `template <AccessEnum Access_t, typename Field_t> bool addField(FieldId<Access_t, Field_t> &field_id, std::string const &field_name);`
   * `template <typename Field_t> bool addField(FieldId<ACIDPTR, Field_t> &field_id, std::string const &field_name);`
   But in the example, we call it with just a string?